File: creational_patterns/java/factory_method/FactoryMethod.java
package creational_patterns.java.factory_method;

// Log types as enums
enum LogType{
    ERROR, DEBUG, INFO
}

// ILogger interface for abstaction
interface ILogger{
    void log(String message);
}

// Concrete logger classes
class ErrorLogger implements ILogger{
    @Override
    public void log(String message){
        System.out.println("[ERROR]: " + message);
    }
}

class DebugLogger implements ILogger{
    @Override
    public void log(String message){
        System.out.println("[DEBUG]: " + message);
    }
}

class InfoLogger implements ILogger{
    @Override
    public void log(String message){
        System.out.println("[INFO]: " + message);
    }
}

// Logger factory - creates a logger object based on log type.
// In simple-factory pattern, Factory class used to take care of creation of ILogger Objects.
// Will be creating ILoggerFactory interface/abstract-class and concrete class that implement the interface, will take care of creation process.
// Will be delegating creation process to specific factories inorder to achieve OCP.
interface ILoggerFactory{
    public ILogger createLogger();
}

class InfoLoggerFactory implements ILoggerFactory {
    @Override
    public ILogger createLogger(){
        return new InfoLogger();
    }
}

class DebugLoggerFactory implements ILoggerFactory {
    @Override
    public ILogger createLogger(){
        return new DebugLogger();
    }
}

class ErrorLoggerFactory implements ILoggerFactory {
    @Override
    public ILogger createLogger(){
        return new ErrorLogger();
    }
}

// Client
class FactoryMethod{
    public static void main(String args[]){
        // Specific Loggers
        ILoggerFactory loggerFactory = new InfoLoggerFactory();
        ILogger logger = loggerFactory.createLogger();
        logger.log("Hola..!");

        
        loggerFactory = new DebugLoggerFactory();
        logger = loggerFactory.createLogger();
        logger.log("Hola..!");

        loggerFactory = new ErrorLoggerFactory();
        logger = loggerFactory.createLogger();
        logger.log("Hola..!");
    }
}

File: creational_patterns/c++/factory_method/FactoryMethod.cpp
#include<iostream>
#include<string>
#include<memory>
using namespace std;

class ILogger{
public:
    virtual void log(string message) = 0;
    virtual ~ILogger() {};
};

class ErrorLogger : public ILogger {
public:
    void log(const string message) override {
        cout << "[ERROR]: " << message << endl;
    }
};

class InfoLogger : public ILogger {
public:
    void log(const string message) override {
        cout << "[INFO]: " << message << endl;
    }
};

class DebugLogger : public ILogger {
public:
    void log(const string message) override {
        cout << "[DEBUG]: " << message << endl;
    }
};


class ILoggerFactory {
public:
    virtual unique_ptr<ILogger> create() = 0;
    virtual ~ILoggerFactory(){};
};

class DebugLoggerFactory : public ILoggerFactory {
public:
    unique_ptr<ILogger> create() override {
        return make_unique<DebugLogger>();
    }
};

class ErrorLoggerFactory : public ILoggerFactory {
public:
    unique_ptr<ILogger> create() override {
        return make_unique<ErrorLogger>();
    }
};

class InfoLoggerFactory : public ILoggerFactory {
public:
    unique_ptr<ILogger> create() override {
        return make_unique<InfoLogger>();
    }
};

int main(){
    unique_ptr<ILoggerFactory> lf = make_unique<ErrorLoggerFactory>();
    unique_ptr<ILogger> logger = lf->create();
    logger->log("Vineel");

    lf = make_unique<DebugLoggerFactory>();
    logger = lf->create();
    logger->log("Vineel");
}